src/main/java/com/shapesecurity/bandolier/es2017/transformations/DeadCodeElimination.java

```diff
@@ -0,0 +1,130 @@
+			}
+			return Pair.of(removedImportSet, ImmutableList.of(new ImportNamespace(defaultBinding, importItem.namespaceBinding, importItem.moduleSpecifier)));
+		} else if (item instanceof FunctionDeclaration) {
+			FunctionDeclaration declaration = (FunctionDeclaration) item;
+			checkName(declaration.name.name);
+			Variable variable = lookup.findVariableDeclaredBy(declaration.name).fromJust();
+			if (variable.references.length == 0) {
+				return Pair.of(removedImportSet, ImmutableList.empty());
+			}
+		} else if (item instanceof ClassDeclaration) {
+			ClassDeclaration declaration = (ClassDeclaration) item;
+			checkName(declaration.name.name);
+			Variable variable = lookup.findVariableDeclaredBy(declaration.name).fromJust();
+			if (variable.references.length == 0) {
+				return Pair.of(removedImportSet, ImmutableList.empty());
+			}
+		} else if (item instanceof VariableDeclarationStatement) {
+			//todo: array/object bindings?
+			ImmutableList<VariableDeclarator> declarators = ((VariableDeclarationStatement) item).declaration.declarators.filter(declarator -> {
+				if (declarator.init.isNothing() || declarator.init.fromJust() instanceof FunctionExpression|| declarator.init.fromJust() instanceof ClassExpression || declarator.init.fromJust() instanceof ArrowExpression) {
+					return true;
+				}
+				if (declarator.binding instanceof BindingIdentifier) {
+					checkName(((BindingIdentifier) declarator.binding).name);
+					Variable variable = lookup.findVariableDeclaredBy((BindingIdentifier) declarator.binding).fromJust();
+					ImmutableSet<Reference> references = variable.references.filter(reference -> reference.node != declarator.binding).uniqByIdentity();
+					return references.length() > 0;
+				}
+				return true;
+			});
+			if (declarators.length == 0) {
+				return Pair.of(removedImportSet, ImmutableList.empty());
+			}
+			return Pair.of(removedImportSet, ImmutableList.of(new VariableDeclarationStatement(new VariableDeclaration(((VariableDeclarationStatement) item).declaration.kind, declarators))));
+		}
+		return Pair.of(removedImportSet, ImmutableList.of(item));
+	}
+
+	public static Pair<ImmutableSet<Pair<String, String>>, Module> removeAllUnusedDeclarations(@Nonnull Module module) {
+		DeadCodeElimination deadCodeElimination = new DeadCodeElimination(new ScopeLookup(ScopeAnalyzer.analyze(module)));
+		ImmutableList<Directive> directives = module.directives;
+		ImmutableList<ImportDeclarationExportDeclarationStatement> statements = module.items;
+		if (module.items.length == 1) { // IIFE
+			ImportDeclarationExportDeclarationStatement item = module.items.maybeHead().fromJust();
+			if (item instanceof CallExpression && ((CallExpression) item).arguments.length == 0 && ((CallExpression) item).callee instanceof FunctionExpression) {
+				directives = module.directives.uniqByEquality().putAll(((FunctionExpression) ((CallExpression) item).callee).body.directives).toList();
```

Comment:
Save a reference to `((FunctionExpression) ((CallExpression) item).callee).body` to make this easier to read.